### ISSUE TITLE
Make layer and source ownership consistent

### ISSUE THREAD

**Issue by ivovandongen** [mapbox/mapbox-gl-native#6995]
Like iOS in https://github.com/mapbox/mapbox-gl-native/issues/6254, we need some refactoring to make the layer/source ownership more convenient for developers.  Atm, after adding a newly created layer/source to the map, the original java reference becomes unusable. To facilitate this a bit, we throw a Java exception to report misuse of the api to the developer. But this is not well understood.

Instead, we can also keep a raw pointer to the layer/source after transferring ownership and make sure the reference to the map is set.

ToDo:
- [x] Get rid of explicit checks for validity
- [x] Refactor core object transfer so that the native map reference is set when adding the layer/source to the map (needed for `CustomLayer::update`)
- [x] Ensure that adding a layer/source to the map that was already added throws a nice Java exception
- [x] Add JavaDoc:
    - `MapboxMap#addLayer`
    - `MapboxMap#removeLayer`
    - `MapboxMap#addSource`
    - `MapboxMap#removeSource`

**Comment by jfirebaugh** [comment]
https://github.com/mapbox/mapbox-gl-native/issues/6959 and the discussion there is also relevant.

**Comment by boundsj** [comment]
I proposed https://github.com/mapbox/mapbox-gl-native/pull/7014 for #6959. If there is interest in using in for Android see https://github.com/mapbox/mapbox-gl-native/pull/7014#issuecomment-259843817

**Comment by ivovandongen** [comment]
@jfirebaugh @boundsj Thanks, I missed that ticket completely.

@boundsj If you could get your PR merged, I'll make the necessary changes to this one to take advantage of it for Android.

**Comment by ivovandongen** [comment]
@jfirebaugh @boundsj  I've created a separate ticket (#7023) for the ownership transfer on remove so these changed are not held up by that.

cc @tobrun @zugaldia
